checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -1,146 +1,763 @@
+   *       met, the field is added to the {@link #releasedFields} set.
+   *   <li>It calls {@link #checkIndirectCalls} to inspect the method represented by the given
+   *       MethodInvocationNode for any indirect calls within it. The method analyzes the
+   *       called-methods set of each argument after the call and computes the @Owning annotation to
+   *       the field or parameter passed as an argument to this call.
+   * </ul>
+   *
+   * @param obligations the set of obligations to search in
+   * @param invocation the MethodInvocationNode
+   */
+  private void checkMethodInvocation(Set<Obligation> obligations, MethodInvocationNode invocation) {
+    if (methodElt != null) {
+      List<? extends VariableTree> paramsOfCurrentMethod = methodTree.getParameters();
+      addOwningParamsFromMethodCall(obligations, invocation, paramsOfCurrentMethod);
+      addOwningReceiverFromMethodCall(obligations, invocation, paramsOfCurrentMethod);
+      checkIndirectCalls(obligations, invocation, paramsOfCurrentMethod);
+    }
+  }
 
-      // This assumes that any MustCall annotation has at most one element.
+  /**
+   * Checks if a MustCall obligation is satisfied via the given method call. A MustCall obligation
+   * of an element is satisfied if the called-methods set contains the target of its must-call
+   * obligation.
+   *
+   * @param invocation the method invocation node being checked for satisfaction of the MustCall
+   *     obligation
+   * @param varElt the variable annotated with the MustCall annotation
+   * @param target the target of the MustCall obligation
+   * @return {@code true} if the MustCall obligation is satisfied
+   */
+  private boolean mustCallObligationSatisfied(
+      MethodInvocationNode invocation, Element varElt, JavaExpression target) {
+
+    List<String> mustCallValues = typeFactory.getMustCallValue(varElt);
+    if (mustCallValues.size() != 1) {
       // TODO: generalize this to MustCall annotations with more than one element.
-      if (mustCallValues.size() == 1
-          && mustCallValues.contains(method.getSimpleName().toString())) {
-        owningFields.add((VariableElement) receiverEl);
+      return false;
+    }
+
+    AccumulationStore cmStoreAfter = typeFactory.getStoreAfter(invocation);
+    @Nullable AccumulationValue cmValue =
+        cmStoreAfter == null ? null : cmStoreAfter.getValue(target);
+    AnnotationMirror cmAnno = null;
+
+    if (cmValue != null) { // When store contains the lhs
+      Set<String> accumulatedValues = cmValue.getAccumulatedValues();
+      if (accumulatedValues != null) { // type variable or wildcard type
+        cmAnno = typeFactory.createCalledMethods(accumulatedValues.toArray(new String[0]));
+      } else {
+        for (AnnotationMirror anno : cmValue.getAnnotations()) {
+          if (AnnotationUtils.areSameByName(
+              anno, "org.checkerframework.checker.calledmethods.qual.CalledMethods")) {
+            cmAnno = anno;
```

Comment:
Should you break out of the `for` loop once `cmAnno` is assigned?  Further computation seems unnecessary.